src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/commissioner/thread/CommissionerUtils.java

```diff
@@ -0,0 +1,35 @@
+package com.google.chip.chiptool.commissioner.thread;
+
+import io.openthread.commissioner.ByteArray;
+
+public class CommissionerUtils {
+
+  public static byte[] getByteArray(ByteArray byteArray) {
+    byte[] ret = new byte[byteArray.size()];
+    for (int i = 0; i < byteArray.size(); ++i) {
+      ret[i] = (byte) byteArray.get(i).shortValue();
+    }
+    return ret;
+  }
+
+  public static ByteArray getByteArray(byte[] byteArray) {
+    ByteArray ret = new ByteArray();
+    for (byte b : byteArray) {
+      ret.add((short) (((short) b) & 0xff));
```

Comment:
Could you explain the cast? Why is the &0xFF not sufficient? Also if we add a byte into a byte array, why even do the 0xFF?

This method could really use some doccomment.

Reply:
@andy31415 I added [documentation](https://github.com/project-chip/connectedhomeip/pull/2492/files#diff-c17c550a573d7e134586c079088ddc6cR33-R41) for this method, please review.

https://github.com/project-chip/connectedhomeip/pull/2492/files#diff-c17c550a573d7e134586c079088ddc6cR33-R41

Reply:
This looks strange to me too. Maybe we should do:

` ret.add(((short) b) & ((short) 0xff));`

Also, having to do this seems like a bug or at least a wrong design choice somewhere. Can we make swig do something sane (at least add a TODO saying this is not how we want it to work)?

e.g. Can we have avoid this by SWIGing native methods that use `char` or `int8_t` not `uint8_t` ?

It is legal in C++ to alias unsigned types with signed ones, and it's also legal to alias any type with the character type.

Reply:
@mspang 
> This looks strange to me too. Maybe we should do:
ret.add(((short) b) & ((short) 0xff));

Agree that your version is better. Will change.

> Also, having to do this seems like a bug or at least a wrong design choice somewhere. Can we make swig do something sane (at least add a TODO saying this is not how we want it to work)?

It makes sense to me that SWIG maps C++ `uint8_t` to Java `short` because Java `byte` cannot hold a C++ `uint8_t` value bigger than 127. SWIG uses only the least-significant byte of the `short` value, so we don't really need `& ((short) 0xff)` at here. I can remove it if you prefer.

> e.g. Can we have avoid this by SWIGing native methods that use char or int8_t not uint8_t ?
It is legal in C++ to alias unsigned types with signed ones, and it's also legal to alias any type with the character type.

`ByteArray` here maps to `std::vector<uint8_t>` in the OT Commissioner C++ library. If you are suggesting we replace it with `std::vector<int8_t>`, it would be a dramatic change and there are significant difficulties:
1. 3rd-party libs (e.g. mbedtls, COSE, cn-cbor...) in OT commissioner are all using `uint8_t *` to represent a byte array, it requires casts everywhere if we use `std::vector<int8_t>`  in OT commissioner.
2. OpenThread projects are all using `uint8_t *` to represent a byte array and I think we tend to keep it consistent among OpenThread projects.

It is doable to tell SWIG to map `uint8_t` to Java `byte` without changing the C++ code. But there are concerns:
1. It is okay to map `uint8_t` to Java `byte` when `uint8_t` standards for a byte, but there are cases that we use `uint8_t` for 8-bits unsigned integers.
2. What about `uint16_t`, `uint32_t` and `uint64_t`? Should we also may them to Java `short`, `int` and `long`? If not, we are making `uint8_t` so weird...
3. Another option is to define a new C++ type `Byte` and have it maps Java `byte`. I tried this but it cannot work with `typedef uint8_t Byte;`. A new `struct` or `class` is required and this will be a dramatic change.

I am not very sure about the best practices of integer conversions between C++ and Java, I'd like to learn more suggests if you have.

Reply:
> @mspang
> 
> > This looks strange to me too. Maybe we should do:
> > ret.add(((short) b) & ((short) 0xff));
> 
> Agree that your version is better. Will change.
> 
> > Also, having to do this seems like a bug or at least a wrong design choice somewhere. Can we make swig do something sane (at least add a TODO saying this is not how we want it to work)?
> 
> It makes sense to me that SWIG maps C++ `uint8_t` to Java `short` because Java `byte` cannot hold a C++ `uint8_t` value bigger than 127. SWIG uses only the least-significant byte of the `short` value, so we don't really need `& ((short) 0xff)` at here. I can remove it if you prefer.
> 
> > e.g. Can we have avoid this by SWIGing native methods that use char or int8_t not uint8_t ?
> > It is legal in C++ to alias unsigned types with signed ones, and it's also legal to alias any type with the character type.
> 
> `ByteArray` here maps to `std::vector<uint8_t>` in the OT Commissioner C++ library. If you are suggesting we replace it with `std::vector<int8_t>`, it would be a dramatic change and there are significant difficulties:

No, I wouldn't suggest changing `std::vector<uint8_t>` to `std::vector<int8_t>` everywhere for Java's sake. My suggestion was just to `reinterpret_cast` from `uint8_t*` to `int8_t*` so that SWIG sees a signed type and does the right thing, instead of us having to undo the widening conversion.

Concretely, I don't see why you can't define a type that's equivalent to `std::span<int8_t>` and initialize it with the vector data, even if the vector data's dynamic type is `std::vector<uint8_t>`. It's perfectly legal in C++ to do this cast; you don't have to change the vector's type to `std::vector<int8_t>` first.

> 
> 1. 3rd-party libs (e.g. mbedtls, COSE, cn-cbor...) in OT commissioner are all using `uint8_t *` to represent a byte array, it requires casts everywhere if we use `std::vector<int8_t>`  in OT commissioner.

Wouldn't it only require casts in the code wrapped using SWIG? I'd think it'd be about the same amount of code than defining these conversions.

> 2. OpenThread projects are all using `uint8_t *` to represent a byte array and I think we tend to keep it consistent among OpenThread projects.
> 
> It is doable to tell SWIG to map `uint8_t` to Java `byte` without changing the C++ code. But there are concerns:
> 
> 1. It is okay to map `uint8_t` to Java `byte` when `uint8_t` standards for a byte, but there are cases that we use `uint8_t` for 8-bits unsigned integers.
> 2. What about `uint16_t`, `uint32_t` and `uint64_t`? Should we also may them to Java `short`, `int` and `long`? If not, we are making `uint8_t` so weird...
> 3. Another option is to define a new C++ type `Byte` and have it maps Java `byte`. I tried this but it cannot work with `typedef uint8_t Byte;`. A new `struct` or `class` is required and this will be a dramatic change.
> 
> I am not very sure about the best practices of integer conversions between C++ and Java, I'd like to learn more suggests if you have.

Reply:
@mspang 
> Concretely, I don't see why you can't define a type that's equivalent to std::span<int8_t> and initialize it with the vector data, even if the vector data's dynamic type is std::vector<uint8_t>. It's perfectly legal in C++ to do this cast; you don't have to change the vector's type to std::vector<int8_t> first.
>
>>3rd-party libs (e.g. mbedtls, COSE, cn-cbor...) in OT commissioner are all using uint8_t * to represent a byte array, it requires casts everywhere if we use std::vector<int8_t> in OT commissioner.
>
>Wouldn't it only require casts in the code wrapped using SWIG? I'd think it'd be about the same amount of code than defining these conversions.

Thanks for your suggestions! The background is that we don't write the C++ wrapper code by ourselves in OT Commissioner. Both the C++ wrapper and the Java code are generated by SWIG based on the OT Commissioner public headers. It is easy to tell SWIG to cast `signed char` to/from `uint8_t` in the generated C++ wrapper code. But this will take effect for all references to `uint8_t` and that is not what we want. I am trying to tell SWIG to only cast `std::vector<signed char>` to/from `std::vector<uint8_t>` but this doesn't seem to be trivial. Will back to you when I have some progress.

Reply:
@mspang I figured out the trick to do this with SWIG and it now maps Java `byte` to/from `uint8_t` for only `std::vector<uint8_t>` (see the PR). We can now get ride of this function because `ByteArray` has a constructor accepts `byte[]`. Please review.